rclcpp/test/benchmark/benchmark_executor.cpp

```diff
@@ -58,65 +58,277 @@ class PerformanceTestExecutor : public PerformanceTest
     rclcpp::shutdown();
   }
 
+  template<class Executer>
+  void executor_spin_some(benchmark::State & st)
+  {
+    Executer executor;
+    for (unsigned int i = 0u; i < kNumberOfNodes; i++) {
+      executor.add_node(nodes[i]);
+      publishers[i]->publish(empty_msgs);
+      executor.spin_some(100ms);
+    }
+
+    callback_count = 0;
+    reset_heap_counters();
+
+    for (auto _ : st) {
+      (void)_;
+      st.PauseTiming();
+      for (unsigned int i = 0u; i < kNumberOfNodes; i++) {
+        publishers[i]->publish(empty_msgs);
+      }
+      st.ResumeTiming();
+
+      callback_count = 0;
+      while (callback_count < kNumberOfNodes) {
+        executor.spin_some(100ms);
+      }
+    }
+    if (callback_count == 0) {
+      st.SkipWithError("No message was received");
+    }
+  }
+
+  template<class Executer>
+  void benchmark_wait_for_work(benchmark::State & st)
+  {
+    class ExecuterDerived : public Executer
+    {
+public:
+      void call_wait_for_work(std::chrono::nanoseconds timeout)
+      {
+        Executer::wait_for_work(timeout);
+      }
+    };
+
+    ExecuterDerived executor;
+    for (unsigned int i = 0u; i < kNumberOfNodes; i++) {
+      executor.add_node(nodes[i]);
+      executor.spin_some(100ms);
+    }
+    // we need one ready event
+    publishers[0]->publish(empty_msgs);
+
+    reset_heap_counters();
+
+    for (auto _ : st) {
+      (void)_;
+
+      st.PauseTiming();
+      // we need one ready event
+      publishers[0]->publish(empty_msgs);
+      st.ResumeTiming();
+
+
+      executor.call_wait_for_work(100ms);
+      st.PauseTiming();
+      executor.spin_some(100ms);
+      st.ResumeTiming();
+    }
+  }
+
   test_msgs::msg::Empty empty_msgs;
   std::vector<rclcpp::Node::SharedPtr> nodes;
   std::vector<rclcpp::Publisher<test_msgs::msg::Empty>::SharedPtr> publishers;
   std::vector<rclcpp::Subscription<test_msgs::msg::Empty>::SharedPtr> subscriptions;
-  int callback_count;
+  uint callback_count;
 };
 
 BENCHMARK_F(PerformanceTestExecutor, single_thread_executor_spin_some)(benchmark::State & st)
 {
-  rclcpp::executors::SingleThreadedExecutor executor;
-  for (unsigned int i = 0u; i < kNumberOfNodes; i++) {
-    executor.add_node(nodes[i]);
-    publishers[i]->publish(empty_msgs);
-    executor.spin_some(100ms);
-  }
+  executor_spin_some<rclcpp::executors::SingleThreadedExecutor>(st);
+}
 
-  callback_count = 0;
-  reset_heap_counters();
+BENCHMARK_F(PerformanceTestExecutor, multi_thread_executor_spin_some)(benchmark::State & st)
+{
+  executor_spin_some<rclcpp::executors::MultiThreadedExecutor>(st);
+}
 
-  for (auto _ : st) {
```

Comment:
maybe 
```cpp
for ([[maybe_unused]] auto _ : st) {
    st.PauseTiming();
``` 
is more readable

Reply:
I agree, but this is not the style used in the rest of the code.